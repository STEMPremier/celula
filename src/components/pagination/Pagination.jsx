/* Copyright 2020 Tallo Inc.,
 *
 * This file is part of Celula.
 *
 * Celula is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
/* eslint-disable react/no-array-index-key */
/* eslint-disable no-plusplus */
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import './pagination.less';

import { SystemIcon } from '../icon';

/**
 * Pagination is the process of dividing data into discrete pages. This graphical control element is frequently used as a navigational aid.
 */
const Pagination = ({
  className,
  canNextPage,
  canPreviousPage,
  currentPage,
  gotoPage,
  nextPage,
  pageCount,
  prevPage,
}) => {
  const classes = cx('ce-pagination', className);

  const renderNumbers = () => {
    const elements = [];

    if (pageCount > 5) {
      for (let i = 1; i < 6; i++) {
        let classList = 'ce-pagination__number';
        let label = '...';
        let navToPage;

        switch (i) {
          case 1:
            label = 1;
            navToPage = 1;

            if (currentPage === 1) classList = `${classList} current-page`;
            break;
          case 2:
            if (currentPage <= 3) {
              label = 2;
              navToPage = 2;
            }

            if (currentPage === 2) {
              classList = `${classList} current-page`;
            }

            if (currentPage > 3) {
              label = '...';
              navToPage = currentPage - 1;
            }
            break;
          case 3:
            if (currentPage <= 3) {
              label = 3;
              navToPage = 3;
            }

            if (currentPage === 3) classList = `${classList} current-page`;

            if (currentPage > 3 && currentPage < pageCount - 1) {
              classList = `${classList} current-page`;
              label = currentPage;
              navToPage = currentPage;
            }

            if (currentPage === pageCount - 1 || currentPage === pageCount) {
              label = pageCount - 2;
              navToPage = pageCount - 2;
            }
            break;
          case 4:
            if (currentPage < pageCount - 2) {
              label = '...';
              navToPage = currentPage + 1;
            }

            if (currentPage === pageCount - 2) {
              label = currentPage + 1;
              navToPage = currentPage + 1;
            }

            if (currentPage === pageCount - 1) {
              classList = `${classList} current-page`;
              label = currentPage;
              navToPage = currentPage;
            }

            if (currentPage === pageCount) {
              label = pageCount - 1;
              navToPage = pageCount - 1;
            }
            break;
          case 5: // extreme right
            label = pageCount;
            navToPage = pageCount;

            if (currentPage === pageCount) {
              classList = `${classList} current-page`;
            }
            break;
          default:
        }

        elements.push(
          <li className={classList} key={i}>
            <button
              type="button"
              onClick={() => gotoPage(navToPage) /* The pages are 0-indexed */}
            >
              {label}
            </button>
          </li>,
        );
      }
    } else {
      for (let i = 1; i <= pageCount; i++) {
        elements.push(
          <li
            className={`ce-pagination__number${
              currentPage === i ? ' current-page' : ''
            }`}
            key={i}
          >
            <button type="button" onClick={() => gotoPage(i)}>
              {i}
            </button>
          </li>,
        );
      }
    }

    return elements;
  };

  return (
    <nav className={classes}>
      <ul className="ce-pagination__container">
        <li
          className={`ce-pagination__arrow${
            !canPreviousPage ? ' ce-arrow--disabled' : ''
          }`}
        >
          <button disabled={!canPreviousPage} onClick={prevPage} type="button">
            <SystemIcon name="navigate" color="black" />
          </button>
        </li>
        {renderNumbers()}
        <li
          className={`ce-pagination__arrow${
            !canNextPage ? ' ce-arrow--disabled' : ''
          }`}
        >
          <button disabled={!canNextPage} onClick={nextPage} type="button">
            <SystemIcon name="navigate" color="black" />
          </button>
        </li>
      </ul>
    </nav>
  );
};

Pagination.propTypes = {
  /**
   * Disable the next page button if there are no more pages.
   */
  canNextPage: PropTypes.bool,
  /**
   * Disable the previous page if there are no prior pages.
   */
  canPreviousPage: PropTypes.bool,
  /**
   * A class name, or string of class names, to add to the `<Pagination />`.
   */
  className: PropTypes.string,
  /**
   * The current page number.
   *
   * NOTE: Due to interactions between `<Pagination />` and `<Table />` this is 0-indexed.
   */
  currentPage: PropTypes.number.isRequired,
  /**
   * A function to trigger to navigate to an arbitrary page.
   */
  gotoPage: PropTypes.func.isRequired,
  /**
   * A function to trigger to navigate to the next page.
   */
  nextPage: PropTypes.func.isRequired,
  /**
   * The total number of pages.
   */
  pageCount: PropTypes.number.isRequired,
  /**
   * A function to trigger to navigate to the previous page.
   */
  prevPage: PropTypes.func.isRequired,
};

Pagination.defaultProps = {
  canNextPage: true,
  canPreviousPage: true,
  className: '',
};

export default Pagination;
/* eslint-enable react/no-array-index-key */
