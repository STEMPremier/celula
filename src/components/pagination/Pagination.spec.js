import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Pagination from '.';

describe('<Pagination />', () => {
  describe('render()', () => {
    test('renders the component', () => {
      const canNextPage = true;
      const canPreviousPage = true;

      const wrapper = shallow(
        <Pagination
          canNextPage={canNextPage}
          canPreviousPage={canPreviousPage}
          currentPage={1}
          gotoPage={() => console.log('Navigate to a different page.')}
          nextPage={() => console.log('Navigate up a page.')}
          pageCount={10}
          prevPage={() => console.log('Navigate down a page.')}
        />,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
