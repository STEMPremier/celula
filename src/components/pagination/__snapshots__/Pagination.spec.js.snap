// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> render() renders the component 1`] = `
<nav
  className="ce-pagination"
>
  <ul
    className="ce-pagination__container"
  >
    <li
      className="ce-pagination__arrow"
    >
      <button
        aria-label="Previous Page"
        disabled={false}
        onClick={[Function]}
        title="Previous Page"
        type="button"
      >
        <SystemIcon
          className=""
          color="black"
          name="navigate"
          size="large"
        />
      </button>
    </li>
    <li
      className="ce-pagination__number current-page"
      key="1"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        1
      </button>
    </li>
    <li
      className="ce-pagination__number"
      key="2"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        2
      </button>
    </li>
    <li
      className="ce-pagination__number"
      key="3"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        3
      </button>
    </li>
    <li
      className="ce-pagination__number"
      key="4"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        ...
      </button>
    </li>
    <li
      className="ce-pagination__number"
      key="5"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        10
      </button>
    </li>
    <li
      className="ce-pagination__arrow"
    >
      <button
        aria-label="Next Page"
        disabled={false}
        onClick={[Function]}
        title="Next Page"
        type="button"
      >
        <SystemIcon
          className=""
          color="black"
          name="navigate"
          size="large"
        />
      </button>
    </li>
  </ul>
</nav>
`;
